--- bumblebee-3.0.1/src/bbsecondary.c.orig	2012-07-26 18:24:26.000000000 +0300
+++ bumblebee-3.0.1/src/bbsecondary.c	2012-08-17 14:38:10.400305127 +0300
@@ -146,6 +146,7 @@
       XORG_BINARY,
       bb_config.x_display,
       "-config", x_conf_file,
+      "-configdir", "/etc/bumblebee/xorg.conf.nvidia.d",
       "-sharevts",
       "-nolisten", "tcp",
       "-noreset",
@@ -156,7 +157,7 @@
       NULL
     };
     if (!*bb_config.mod_path) {
-      x_argv[12] = 0; //remove -modulepath if not set
+      x_argv[14] = 0; //remove -modulepath if not set
     }
     //close any previous pipe, if it (still) exists
     if (bb_status.x_pipe[0] != -1){close(bb_status.x_pipe[0]); bb_status.x_pipe[0] = -1;}
