diff -ur lsd2dsl-0.5.6.orig/lib/lsd/AbbreviationDictionaryDecoder.cpp lsd2dsl-0.5.6/lib/lsd/AbbreviationDictionaryDecoder.cpp
--- lsd2dsl-0.5.6.orig/lib/lsd/AbbreviationDictionaryDecoder.cpp	2023-03-19 13:38:29.000000000 +0100
+++ lsd2dsl-0.5.6/lib/lsd/AbbreviationDictionaryDecoder.cpp	2024-02-19 09:28:43.254842564 +0100
@@ -4,6 +4,8 @@
 #include "UserDictionaryDecoder.h"
 #include "tools.h"
 
+#include <cassert>
+
 namespace dictlsd {
 
 std::u16string readXoredPrefix(IBitStream* bstr, int len) {
diff -ur lsd2dsl-0.5.6.orig/lib/lsd/LSDOverlayReader.cpp lsd2dsl-0.5.6/lib/lsd/LSDOverlayReader.cpp
--- lsd2dsl-0.5.6.orig/lib/lsd/LSDOverlayReader.cpp	2023-03-19 13:38:29.000000000 +0100
+++ lsd2dsl-0.5.6/lib/lsd/LSDOverlayReader.cpp	2024-02-19 09:28:19.618073851 +0100
@@ -6,6 +6,7 @@
 
 #include <zlib.h>
 #include <stdexcept>
+#include <cassert>
 
 namespace dictlsd {
 
diff -ur lsd2dsl-0.5.6.orig/lib/lsd/SystemDictionaryDecoder.cpp lsd2dsl-0.5.6/lib/lsd/SystemDictionaryDecoder.cpp
--- lsd2dsl-0.5.6.orig/lib/lsd/SystemDictionaryDecoder.cpp	2023-03-19 13:38:29.000000000 +0100
+++ lsd2dsl-0.5.6/lib/lsd/SystemDictionaryDecoder.cpp	2024-02-19 09:30:51.648728938 +0100
@@ -2,6 +2,8 @@
 #include "BitStream.h"
 #include "tools.h"
 
+#include <cassert>
+
 namespace dictlsd {
 
 SystemDictionaryDecoder::SystemDictionaryDecoder(bool xoring)
diff -ur lsd2dsl-0.5.6.orig/lib/lsd/UserDictionaryDecoder.cpp lsd2dsl-0.5.6/lib/lsd/UserDictionaryDecoder.cpp
--- lsd2dsl-0.5.6.orig/lib/lsd/UserDictionaryDecoder.cpp	2023-03-19 13:38:29.000000000 +0100
+++ lsd2dsl-0.5.6/lib/lsd/UserDictionaryDecoder.cpp	2024-02-19 09:30:29.531967148 +0100
@@ -5,6 +5,7 @@
 #include <stdint.h>
 #include <string.h>
 #include <stdexcept>
+#include <cassert>
 
 namespace dictlsd {
 
