Only in lsd2dsl-0.5.4: .idea
diff -ur lsd2dsl-0.5.4.orig/tests/test-utils.h lsd2dsl-0.5.4/tests/test-utils.h
--- lsd2dsl-0.5.4.orig/tests/test-utils.h	2021-08-16 05:31:10.000000000 +0200
+++ lsd2dsl-0.5.4/tests/test-utils.h	2022-06-17 11:52:16.531632580 +0200
@@ -15,7 +15,9 @@
     auto filesize = ftell(f);
     std::vector<uint8_t> res(filesize);
     fseek(f, 0, SEEK_SET);
-    fread(&res[0], 1, res.size(), f);
+    auto read = fread(&res[0], 1, res.size(), f);
+    if (read != 1)
+        throw std::runtime_error("can't read file");
     fclose(f);
     return res;
 }
