diff -ur bbswitch-0.8.old/bbswitch.c bbswitch-0.8/bbswitch.c
--- bbswitch-0.8.old/bbswitch.c	2017-07-03 10:49:05.078648597 +0200
+++ bbswitch-0.8/bbswitch.c	2013-12-05 00:22:06.019830784 +0100
@@ -31,7 +31,7 @@
 #include <linux/pci.h>
 #include <linux/acpi.h>
 #include <linux/module.h>
-#include <asm/uaccess.h>
+#include <linux/uaccess.h>
 #include <linux/suspend.h>
 #include <linux/seq_file.h>
 #include <linux/pm_runtime.h>
