diff -ur evdi-90e9679591daf2d1d57bc8d7fa7d717cce62c933.orig/module/evdi_gem.c evdi-90e9679591daf2d1d57bc8d7fa7d717cce62c933/module/evdi_gem.c
--- evdi-90e9679591daf2d1d57bc8d7fa7d717cce62c933.orig/module/evdi_gem.c	2020-11-03 09:31:03.000000000 +0100
+++ evdi-90e9679591daf2d1d57bc8d7fa7d717cce62c933/module/evdi_gem.c	2020-11-06 11:25:09.108524468 +0100
@@ -306,6 +306,10 @@
 {
 	struct evdi_gem_object *bo = to_evdi_bo(obj);
 
+#if KERNEL_VERSION(5, 10, 0) <= LINUX_VERSION_CODE
+	return drm_prime_pages_to_sg(obj->dev, bo->pages, bo->base.size >> PAGE_SHIFT);
+#else
 	return drm_prime_pages_to_sg(bo->pages, bo->base.size >> PAGE_SHIFT);
+#endif
 }
 
