From 9b857fc42cdc4921e1e3f190c5ea907774e04758 Mon Sep 17 00:00:00 2001
From: Nikias Bassen <nikias@gmx.li>
Date: Wed, 20 Nov 2019 01:40:08 +0100
Subject: [PATCH] tools: Fix ideviceinfo -s option

---
 tools/ideviceinfo.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/tools/ideviceinfo.c b/tools/ideviceinfo.c
index bd3819e2..f52cd005 100644
--- a/tools/ideviceinfo.c
+++ b/tools/ideviceinfo.c
@@ -138,6 +138,7 @@ int main(int argc, char *argv[])
 		{ "network", no_argument, NULL, 'n' },
 		{ "domain", required_argument, NULL, 'q' },
 		{ "key", required_argument, NULL, 'k' },
+		{ "simple", no_argument, NULL, 's' },
 		{ "xml", no_argument, NULL, 'x' },
 		{ NULL, 0, NULL, 0}
 	};
@@ -146,7 +147,7 @@ int main(int argc, char *argv[])
 	signal(SIGPIPE, SIG_IGN);
 #endif
 
-	while ((c = getopt_long(argc, argv, "dhu:nq:k:x", longopts, NULL)) != -1) {
+	while ((c = getopt_long(argc, argv, "dhu:nq:k:sx", longopts, NULL)) != -1) {
 		switch (c) {
 		case 'd':
 			idevice_set_debug_level(1);
