diff -ur openssh-6.4p1.orig/pathnames.h openssh-6.4p1/pathnames.h
--- openssh-6.4p1.orig/pathnames.h	2014-01-03 10:35:46.000000000 +0200
+++ openssh-6.4p1/pathnames.h	2014-01-03 10:37:17.764277188 +0200
@@ -58,7 +58,7 @@
  * The directory in user's home directory in which the files reside. The
  * directory should be world-readable (though not all files are).
  */
-#define _PATH_SSH_USER_DIR		".ssh"
+#define _PATH_SSH_USER_DIR		".config/ssh"
 
 /*
  * Per-user file containing host keys of known hosts.  This file need not be
