diff -ur openssh-6.4p1.orig/contrib/ssh-copy-id openssh-6.4p1/contrib/ssh-copy-id
--- openssh-6.4p1.orig/contrib/ssh-copy-id	2013-06-05 15:48:45.000000000 +0300
+++ openssh-6.4p1/contrib/ssh-copy-id	2014-01-20 15:55:42.404922710 +0200
@@ -56,7 +56,7 @@
   fi
 fi
 
-DEFAULT_PUB_ID_FILE=$(ls -t ${HOME}/.ssh/id*.pub 2>/dev/null | grep -v -- '-cert.pub$' | head -n 1)
+DEFAULT_PUB_ID_FILE=$(ls -t ${XDG_CONFIG_HOME}/ssh/id*.pub 2>/dev/null | grep -v -- '-cert.pub$' | head -n 1)
 
 usage () {
   printf 'Usage: %s [-h|-?|-n] [-i [identity_file]] [-p port] [[-o <ssh -o options>] ...] [user@]hostname\n' "$0" >&2
@@ -198,7 +198,7 @@
   eval set -- "$SSH_OPTS"
 
   umask 0177
-  local L_TMP_ID_FILE=$(mktemp ~/.ssh/ssh-copy-id_id.XXXXXXXXXX)
+  local L_TMP_ID_FILE=$(mktemp ${XDG_CONFIG_HOME}/ssh/ssh-copy-id_id.XXXXXXXXXX)
   if test $? -ne 0 || test "x$L_TMP_ID_FILE" = "x" ; then
     echo "mktemp failed" 1>&2
     exit 1
