diff -ur firewalld-0.7.5.orig/src/firewall/core/ipXtables.py firewalld-0.7.5/src/firewall/core/ipXtables.py
--- firewalld-0.7.5.orig/src/firewall/core/ipXtables.py	2020-06-30 21:41:34.000000000 +0200
+++ firewalld-0.7.5/src/firewall/core/ipXtables.py	2020-08-14 15:54:07.764970601 +0200
@@ -618,9 +618,12 @@
         ret = [ ]
         output = ""
         try:
+            old_command = self._command
+            self._command += "-legacy"
             output = self.__run(["-p",
                                  "icmp" if self.ipv == "ipv4" else "ipv6-icmp",
                                  "--help"])
+            self._command = old_command
         except ValueError as ex:
             if self.ipv == "ipv4":
                 log.debug1("iptables error: %s" % ex)
