diff --git a/src/firewall/core/fw_nm.py b/src/firewall/core/fw_nm.py
index 37282a1a..ec3652df 100644
--- a/src/firewall/core/fw_nm.py
+++ b/src/firewall/core/fw_nm.py
@@ -34,6 +34,10 @@ except ValueError:
     _nm_imported = False
 else:
     try:
+        # if Nftables is created after "from gi.repository import NM" it messes it up and it stops sending
+        # commands properly over netlink and thus not creating any rules and it causes ERROR: 'handle' error
+        import nftables as sysnftables
+        sysnftables.Nftables()
         from gi.repository import NM
         _nm_imported = True
     except (ImportError, ValueError, GLib.Error):
