--- configure.ac.ori	2013-06-05 01:48:58.997813488 +0200
+++ configure.ac	2013-06-05 02:21:12.940046659 +0200
@@ -1,21 +1,17 @@
 AC_INIT([libconcord], [1.0], [bug-automake@gnu.org])
 AM_INIT_AUTOMAKE([-Wall -Werror foreign])
 AC_CONFIG_MACRO_DIR([m4])
-AC_PROG_LIBTOOL
+
+AM_PROG_AR
 AC_PROG_CXX
-a=1
-AC_CHECK_HEADER(usb.h, [], [a=0])
-AC_CHECK_LIB(usb, usb_init, [], [a=0])
-if test $a == 0
-then
-	AC_MSG_ERROR([Error, libusb is missing!])
-fi
-AC_CHECK_HEADER(zzip/lib.h, [], [a=0])
-AC_CHECK_LIB(zzip, zzip_dir_open, [], [a=0])
-if test $a == 0
-then
-	AC_MSG_ERROR([Error, libzzip is missing!])
-fi
+AC_PROG_LIBTOOL
+
+PKG_CHECK_MODULES([libusb], [libusb])
+CPPFLAGS="$CPPFLAGS $libusb_CFLAGS"
+
+PKG_CHECK_MODULES([zziplib], [zziplib])
+CPPFLAGS="$CPPFLAGS $zziplib_CFLAGS"
+
 AC_CONFIG_HEADERS([config.h])
 AC_CONFIG_FILES([
  Makefile
