diff -ur xrandr-1.4.3.orig/xrandr.c xrandr-1.4.3/xrandr.c
--- xrandr-1.4.3.orig/xrandr.c	2014-08-02 08:46:28.000000000 +0300
+++ xrandr-1.4.3/xrandr.c	2015-01-26 14:08:35.892660991 +0200
@@ -1254,7 +1254,7 @@
 	    output->transform.transform.matrix[0][0] = XDoubleToFixed (sx);
 	    output->transform.transform.matrix[1][1] = XDoubleToFixed (sy);
 	    output->transform.transform.matrix[2][2] = XDoubleToFixed (1.0);
-	    if (sx != 1 || sy != 1)
+	    if ((sx != 1 || sy != 1) && (sx != 0.5 || sy != 0.5))
 		output->transform.filter = "bilinear";
 	    else
 		output->transform.filter = "nearest";
