--- obdgpslogger-0.16.orig/src/gui/convertwindow.fl     2010-05-02
01:24:44.000000000 +0200
+++ obdgpslogger-0.16/src/gui/convertwindow.fl  2011-09-18
16:35:23.408642135 +0200
@@ -102,8 +102,6 @@ if(0 < f.count())
     }
     code {mMainui = obdui;} {}
   }
-  Function {~obdconvertwindow()} {open
-  } {}
   decl {OBDUI *mMainui;} {}
   decl {enum obd_convert_type mConvertType;} {}
   Function {show()} {open
