diff -ur libssh-0.6.3.orig/src/options.c libssh-0.6.3/src/options.c
--- libssh-0.6.3.orig/src/options.c	2014-02-04 16:59:24.000000000 +0200
+++ libssh-0.6.3/src/options.c	2014-03-13 14:55:34.261447434 +0200
@@ -526,7 +526,7 @@
             v = value;
             SAFE_FREE(session->opts.sshdir);
             if (v == NULL) {
-                session->opts.sshdir = ssh_path_expand_tilde("~/.ssh");
+                session->opts.sshdir = ssh_path_expand_tilde("~/.config/ssh");
                 if (session->opts.sshdir == NULL) {
                     return -1;
                 }
