--- a/src/runtime/base/preg.cpp	2012-07-19 12:41:59.000000000 +0300
+++ b/src/runtime/base/preg.cpp	2012-07-19 12:49:50.717108953 +0300
@@ -135,7 +135,8 @@
      * We use a quick pcre_info() check to see whether cache is corrupted,
      * and if it is, we flush it and compile the pattern from scratch.
      */
-    if (pcre_info(pce->re, NULL, NULL) == PCRE_ERROR_BADMAGIC) {
+    size_t where;
+    if (pcre_fullinfo(pce->re, NULL, 0, &where) == PCRE_ERROR_BADMAGIC) {
       pcre_cache.cleanup();
     } else {
 #if HAVE_SETLOCALE
