--- /usr/share/laptop-mode-tools/modules/hdparm	2013-05-14 18:00:38.000000000 +0300
+++ /usr/share/laptop-mode-tools/modules/hdparm.new	2013-05-14 17:59:24.000000000 +0300
@@ -41,7 +41,7 @@
 	# It assumes more or less recent udev (> 070)
 	if [ $HAVE_UDEVINFO -ne 0 ] ; then
 		log "VERBOSE" "Querying $1 media type using udevinfo: "
-		if [ -x /sbin/udevadm ]; then
+		if [ -x "$(which udevadm)" ]; then
 			eval "$(udevadm info -q env -n $1 | egrep '(ID_TYPE=|ID_BUS=)' )"
 		else
 			eval "$(udevinfo -q env -n $1 | egrep '(ID_TYPE=|ID_BUS=)' )"
