diff -ur cppast.orig/test/CMakeLists.txt cppast/test/CMakeLists.txt
--- cppast.orig/test/CMakeLists.txt	2019-08-16 12:58:48.000000000 +0200
+++ cppast/test/CMakeLists.txt	2019-09-17 15:56:49.830431877 +0200
@@ -2,12 +2,6 @@
 # This file is subject to the license terms in the LICENSE file
 # found in the top-level directory of this distribution.
 
-# download catch
-if(NOT EXISTS ${CMAKE_CURRENT_BINARY_DIR}/catch.hpp)
-    file(DOWNLOAD https://raw.githubusercontent.com/catchorg/Catch2/master/single_include/catch2/catch.hpp
-            ${CMAKE_CURRENT_BINARY_DIR}/catch.hpp)
-endif()
-
 set(tests
         code_generator.cpp
         cpp_alias_template.cpp
diff -ur cppast.orig/test/cpp_token.cpp cppast/test/cpp_token.cpp
--- cppast.orig/test/cpp_token.cpp	2019-08-16 12:58:48.000000000 +0200
+++ cppast/test/cpp_token.cpp	2019-09-17 15:55:49.060765515 +0200
@@ -4,7 +4,7 @@
 
 #include <cppast/cpp_token.hpp>
 
-#include <catch.hpp>
+#include <catch2/catch.hpp>
 
 #include <algorithm>
 #include <initializer_list>
diff -ur cppast.orig/test/libclang_parser.cpp cppast/test/libclang_parser.cpp
--- cppast.orig/test/libclang_parser.cpp	2019-08-16 12:58:48.000000000 +0200
+++ cppast/test/libclang_parser.cpp	2019-09-17 15:56:22.412131347 +0200
@@ -2,7 +2,7 @@
 // This file is subject to the license terms in the LICENSE file
 // found in the top-level directory of this distribution.
 
-#include <catch.hpp>
+#include <catch2/catch.hpp>
 
 #include <cppast/libclang_parser.hpp>
 
diff -ur cppast.orig/test/parser.cpp cppast/test/parser.cpp
--- cppast.orig/test/parser.cpp	2019-08-16 12:58:48.000000000 +0200
+++ cppast/test/parser.cpp	2019-09-17 15:56:34.345262170 +0200
@@ -4,7 +4,7 @@
 
 #include <cppast/parser.hpp>
 
-#include <catch.hpp>
+#include <catch2/catch.hpp>
 
 using namespace cppast;
 
diff -ur cppast.orig/test/preprocessor.cpp cppast/test/preprocessor.cpp
--- cppast.orig/test/preprocessor.cpp	2019-08-16 12:58:48.000000000 +0200
+++ cppast/test/preprocessor.cpp	2019-09-17 15:56:00.276888579 +0200
@@ -1,4 +1,4 @@
-#include <catch.hpp>
+#include <catch2/catch.hpp>
 #include <fstream>
 
 #include "libclang/preprocessor.hpp"
diff -ur cppast.orig/test/test.cpp cppast/test/test.cpp
--- cppast.orig/test/test.cpp	2019-08-16 12:58:48.000000000 +0200
+++ cppast/test/test.cpp	2019-09-17 15:56:43.178358983 +0200
@@ -3,4 +3,4 @@
 // found in the top-level directory of this distribution.
 
 #define CATCH_CONFIG_MAIN
-#include <catch.hpp>
+#include <catch2/catch.hpp>
diff -ur cppast.orig/test/test_parser.hpp cppast/test/test_parser.hpp
--- cppast.orig/test/test_parser.hpp	2019-08-16 12:58:48.000000000 +0200
+++ cppast/test/test_parser.hpp	2019-09-17 15:56:10.311998655 +0200
@@ -7,7 +7,7 @@
 
 #include <fstream>
 
-#include <catch.hpp>
+#include <catch2/catch.hpp>
 
 #include <cppast/code_generator.hpp>
 #include <cppast/cpp_class.hpp>
