diff -ur ios-webkit-debug-proxy-5875c9bd8dcb9afcefd071affd46a373ea8fcee1.orig/src/webinspector.c ios-webkit-debug-proxy-5875c9bd8dcb9afcefd071affd46a373ea8fcee1/src/webinspector.c
--- ios-webkit-debug-proxy-5875c9bd8dcb9afcefd071affd46a373ea8fcee1.orig/src/webinspector.c	2019-11-14 05:16:53.000000000 +0100
+++ ios-webkit-debug-proxy-5875c9bd8dcb9afcefd071affd46a373ea8fcee1/src/webinspector.c	2019-11-21 15:45:46.022452836 +0100
@@ -53,11 +53,12 @@
 // CONNECT
 //
 
-#ifndef HAVE_IDEVICE_CONNECTION_GET_FD
 // based on libimobiledevice 1.2.0
 enum connection_type {
-  CONNECTION_USBMUXD = 1
+	CONNECTION_USBMUXD = 1
 };
+
+#ifndef HAVE_IDEVICE_CONNECTION_GET_FD
 struct idevice_connection_private {
   char *udid;  // added in v1.1.6
   enum connection_type type;
@@ -92,7 +93,7 @@
 // based on latest libimobiledevice/src/idevice.h
 struct idevice_connection_private {
   idevice_t device;
-  enum idevice_connection_type type;
+  enum connection_type type;
   void *data;
   void *ssl_data;
 };
