diff -ur cndrvcups-capt-2.60-1.orig/statusui/src/Makefile.am cndrvcups-capt-2.60-1/statusui/src/Makefile.am
--- cndrvcups-capt-2.60-1.orig/statusui/src/Makefile.am	2013-01-23 03:43:24.000000000 +0200
+++ cndrvcups-capt-2.60-1/statusui/src/Makefile.am	2014-01-27 16:28:01.791099364 +0200
@@ -42,5 +42,5 @@
 	calibrationsettingsdlg.c calibrationsettingsdlg.h
 
 
-captstatusui_LDADD = @PACKAGE_LIBS@ $(INTLLIBS) -lxml2 -lcups ../cnsktmodule/libcnsktmodule.la
+captstatusui_LDADD = @PACKAGE_LIBS@ $(INTLLIBS) -lxml2 -lcups -lpthread ../cnsktmodule/libcnsktmodule.la
 
diff -ur cndrvcups-capt-2.60-1.orig/statusui/src/ppapdata.c cndrvcups-capt-2.60-1/statusui/src/ppapdata.c
--- cndrvcups-capt-2.60-1.orig/statusui/src/ppapdata.c	2013-06-27 04:18:49.000000000 +0300
+++ cndrvcups-capt-2.60-1/statusui/src/ppapdata.c	2014-01-27 16:27:49.671098784 +0200
@@ -23,6 +23,7 @@
 #include <string.h>
 
 #include <cups/cups.h>
+#include <cups/ppd.h>
 #include "uimain.h"
 #include "cnsktmodule.h"
 
diff -ur cndrvcups-capt-2.60-1.orig/statusui/src/uimain.c cndrvcups-capt-2.60-1/statusui/src/uimain.c
--- cndrvcups-capt-2.60-1.orig/statusui/src/uimain.c	2013-06-27 03:56:24.000000000 +0300
+++ cndrvcups-capt-2.60-1/statusui/src/uimain.c	2014-01-27 16:27:49.671098784 +0200
@@ -28,6 +28,7 @@
 #include "interface.h"
 #include "data_process.h"
 #include <cups/cups.h>
+#include <cups/ppd.h>
 #define	MAX_WORD_SIZE	512
 
 #define LOCALE_STRING_JA	"ja_JP.UTF-8"
